src/server.rs

```diff
@@ -53,11 +55,33 @@ macro_rules! into_service {
     ($into:expr) => {{
         let inner = crate::service($into);
         make_service_fn(move |transport| {
-            let inner = inner.clone();
-            let remote_addr = Transport::remote_addr(transport);
-            future::ok::<_, Infallible>(service_fn(move |req| {
-                inner.call_with_addr(req, remote_addr)
-            }))
+            if Transport::upgrade_to_https(transport) {
+                future::ok::<_, Infallible>(Either::B(service_fn(move |req| {
+                    let response = if let Some(host) = req
+                        .headers()
+                        .get("host")
+                        .and_then(|header| header.to_str().ok())
+                    {
+                        Response::builder()
+                            .status(301)
+                            .header("Location", format!("https://{}{}", host, req.uri()))
+                            .body(Body::default())
+                            .unwrap()
+                    } else {
+                        Response::builder()
+                            .status(403)
```

Comment:
When requesting HTTP, it should always send a `Host` header.
Also currently when a HTTPS connection fails for other reasons, it will just send back nothing, I could make it send this instead? Same for HTTP actually.